compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp

```diff
@@ -96,6 +97,77 @@ static bool getTilingOptionsFromConfig(func::FuncOp funcOp, int64_t tilingLevel,
   return false;
 }
 
+/// Computes the canonical shape used to vectorize this dispatch. Retrieves
+/// the vectorization tile sizes out of the lowering config and adjusts them to
+/// the format expected by the Linalg vectorizer.
+static SmallVector<int64_t> getCanonicalVectorShape(func::FuncOp funcOp) {
+  FailureOr<Operation *> rootOp = getRootOp(funcOp);
+  if (failed(rootOp)) {
+    return {};
+  }
+
+  unsigned numTileLevels =
+      mlir::iree_compiler::getNumTileLevels(rootOp.value());
+  if (numTileLevels < 3) {
+    return {};
+  }
+
+  // Retrieve the tile sizes from the last two tiling levels used for
+  // vectorization.
+  SmallVector<int64_t> canonicalVectorShape =
+      mlir::iree_compiler::getTileSizes(rootOp.value(), numTileLevels - 2);
+  SmallVector<int64_t> reductionTileSizes =
+      mlir::iree_compiler::getTileSizes(rootOp.value(), numTileLevels - 1);
+
+  if (!reductionTileSizes.empty()) {
+    assert(canonicalVectorShape.size() == reductionTileSizes.size() &&
+           "Unexpected tile sizes");
+
+    // Combine reduction tile sizes into the canonical vector shape.
+    for (int i = 0, end = canonicalVectorShape.size(); i < end; ++i) {
+      if (reductionTileSizes[i] > 0)
+        canonicalVectorShape[i] = reductionTileSizes[i];
+    }
+  }
```

Comment:
What's the difference between reductionTileSizes and canonicalVectorShape? Why not use reductionTileSizes directly if you want to assign it back to canonicalVectorShape? (When is reductionTileSize 0 but canonicalVectorShape not 0?). Perhaps you can add a comment about that invariant here.

Reply:
parallel and reduction tile sizes are the last two tiling levels in the lowering config tuples. The parallel one is assigned to the canonical vector shape (line 117, `numTileLevels - 2`) and the loop above is copying the reduction ones. I improved the comment.